### ISSUE TITLE
Captions with streaming libraries

### ISSUE THREAD

**Issue by friday** [sampotts/plyr#994]
I'm collecting all the info I have in one issue (and intend to keep it updated by editing).

Plyr 3.3.10 added a property `captions.update` for cases when you want Plyr to update the language selection if tracks are added after the instance is initialized.

Ex:
```js
new Plyr(video, {captions: {update: true}});
video.addTextTrack('subtitles', 'Volapük', 'vo'); // Should turn up in the menu now
```

`captions.update` is `false` by default, since the tracks created with it enabled will likely not have any cues, and hence selecting it won't "work". You can however make it work by listening to the `languagechange` event and handling the necessary update there.

__Hls.js__
Status: Somewhat unreliable (still need a setTimeout), but mostly working.
Demo: https://jsfiddle.net/nfyq12qc/

__Dash.js__
Status: It works, but Dash does convert "hidden" text tracks back to "showing" (see: https://github.com/Dash-Industry-Forum/dash.js/issues/1746). In our case it seems to mostly be happening in Firefox, which supports hiding the text track via the `::cue` css selector
Demo: https://jsfiddle.net/g0uxL0kq/2/

__Shaka Player__
Shaka Player only creates one TextTrack withouts any other useful information at all. So it's not possibly to use it with Plyr currently, at least if you want to be able to switch languages. You can however add your own html5 `<track>`-elements if your captions are in the VTT format. You also should keep `captions.update` disabled (default) since otherwise it will just create the option "Shaka Player TextTrack".

Left to do:
1. Keep internal reference to track index rather than just language code (which may not be unique and hence impossible to differentiate in the `languagechange` event).
2. Update docs with Plyr 3.3.10 (or newer) and these examples
3. Expose the internal method to set languages for Shaka Player (I think we should wait with this until we have Plugin support and do it that way instead) See #750

**Comment by indiecore-fm** [comment]
Amazing work! For some reason on my end I can see the different caption options for HLS but they are not showing themselves on the player. Dash is working perfectly though!

**Comment by friday** [comment]
Yeah. Have seen that in HLS too occasionally. Like the German track in the demo ^. Not sure what's happening there.

**Comment by friday** [comment]
Updated the hls.js demo to use the hls.js canary version (https://cdn.jsdelivr.net/npm/hls.js@canary) since that seems to fix all the issues I had in the demo (the German captions not working and some captions not showing the first time).

**Comment by gurupras** [comment]
Is there any way we can we get to work in the general case? I'd like to dynamically add textTracks to HTML5 videos. How can I achieve this? Here's an [example](https://codepen.io/gurupras/pen/ZRpdoe)

**Comment by friday** [comment]
This should work:

```js
new Plyr(video, {captions: {update: true}});
const track = video.addTextTrack('subtitles', 'Volapük', 'vo');
track.addCue(new VTTCue(0, 3, 'Your first cue'));
```

Not sure why you're doing it via the dom. It should work too. I haven't tried for this issue, but I have done this before and it created the `textTracks` objects. `plyr.media.textTracks.length` is 0 in your demo, so something isn't working.

**Comment by gurupras** [comment]
> Not sure why you're doing it via the dom.
I was checking if my code works for the native HMTL5 player (it does) and was hence doing it via the DOM. Why is it that I'm able to add a TextTrack and have it work via the native player but not through Plyr?

In my usecase, ideally, I'd like to set it up to use `src` rather than go through each cue and add it manually.

**Comment by friday** [comment]
I see. If you have URLs rather than creating them as blobs I can see the benefit to using the dom. Make sure you actually get this working without Ply first, and that `video.textTracks.length` and `video.children.length` isn't 0. Otherwise it has nothing to do with Plyr.

**Comment by gurupras** [comment]
It does work without Plyr: Here's the same [example](https://codepen.io/gurupras/pen/NzbKGM) with the native player.

**Comment by friday** [comment]
Neither the video nor the caption loads. You have probably installed a browser plugin like "Allow-Control-Allow-Origin: *" or "Cors toggle" install to override cors. 

Also, you need to enable the textTrack:
```js
document.querySelector('#video').textTracks[0].mode = 'showing'; // use 'hidden' instead of 'showing' for Plyr
```

Not sure how to do that with the dom, but probably the `default` property.

**Comment by gurupras** [comment]
right, sorry, I do have that enabled. I will try and host a video and allow
CORS to get this working

**Comment by friday** [comment]
I forked your pen, changed the sources, set default to true, and used `srcLang` (not language). It works: https://codepen.io/fullkornslimpa/pen/ZRBYrR?editors=1011

**Comment by gurupras** [comment]
Perfect! Thanks for taking the time out to resolve this.

**Comment by friday** [comment]
Plyr 3.3.12 was just released, with support for `currentTrack` getter/setter for the track index. In addition to the previous `language` getter/setter. I think that's about what we can do from our side for support with hls.js and dash.js. Shaka Player needs more, but it's outside of the scope of what I think we need to support/handle, and not a priority for me since they don't support changing tracks via the native HTML5 video controls and/or proper `textTracks` (#750)

In addition to this, I noticed dash.js didn't need to handle the `languagechange` event at all. I just assumed it did, coming from hls.js.

The Plyr documentation has been upgraded with updated examples (same ones as in here)

We could still use some kind of compatibility table in the wiki or main docs I think. Different providers or libraries support for captions, handling quality and other features.
